recipes/ebs.rb

```diff
@@ -23,26 +23,39 @@
   creds = data_bag_item('secrets', 'aws_credentials')['Storage']
 end
 
-if nvme_instance?
-  log 'Due to how EC2 supports EBS volumes on some NVMe instances, we cannot mount them at this time.'
-else
-  node['storage']['ebs_volumes'].each_with_index do |(name, conf), i|
-    aws_ebs_volume name do
-      conf.each { |key, value| send(key, value) }
-      if creds
-        aws_access_key creds['access_key_id']
-        aws_secret_access_key creds['secret_access_key']
-      end
-      action %i(create attach)
+node['storage']['ebs_volumes'].each_with_index do |(name, conf), i|
+  aws_ebs_volume name do
+    conf.each { |key, value| send(key, value) }
+    if creds
+      aws_access_key creds['access_key_id']
+      aws_secret_access_key creds['secret_access_key']
     end
+    action %i(create attach)
+  end
 
-    mount_point = "/mnt/ebs#{i}"
-    device_name = conf['device']
-
-    storage_format_mount mount_point do
-      device_name device_name
-    end
+  mount_point = "/mnt/ebs#{i}"
+  device_name = if nvme_instance?
+                  if node['storage'].attribute?('ebs_mounts') &&
+                     node['storage']['ebs_mounts'].include?('mount_point')
+                    Chef::Log.info "#{mount_point} already mounted."
+                    node['filesystem']['by_mountpoint'][mount_point]['devices'].first
+                  else
+                    # Get the list of filesystem devices, sort them in order,
+                    # grab the last one, and increment the device number.
+                    # This ensures that any EBS volumes are assigned an open device number.
+                    node['filesystem']['by_device'].keys.grep(%r{/dev/nvme}).sort.last
```

Comment:
This worries me a bit, since alphanumeric sort isn't the same as a numeric sort:

```
> "0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21".split(" ").sort().join(" ")
'0 1 10 11 12 13 14 15 16 17 18 19 2 20 21 3 4 5 6 7 8 9'
```

Do we need to do something more clever here?

Reply:
Yeah, this will definitely fall down if we mount >= 10 NVMe devices, for sure.

It’s probably safe as-is, since that many NVMe + EBS devices is highly unlikely.

In order to get these sorted properly, however, it’s a bit of a chose, b/c here’s an example array of values:

```ruby
chef (13.10.0)> node['filesystem']['by_device'].keys.grep(%r{/dev/nvme})
 => ["/dev/nvme0n1p1", "/dev/nvme0n1"]
```

As a result, to sort naturally, we would need to:

1. Filter out any “part” devices (e.g., the `p1` above)
1. Extract the device number (the first digit)
2. Sort an array of those numbers
3. Rebuild an array of strings by adding back the full path (which, thankfully, can be assumed to always be `%r{/dev/nvme(\d)p1}`

That’s pretty torturous to do, so perhaps not worth the effort for a decidedly unlikely event.

Reply:
I agree it's not worth the effort to sort out the full fix.

I recommend at least a comment here just for future reference, though! (If having 10 disks was ever likely, I'd say we should maybe throw an error or omit devices beyond some limit, but I think it'll never come up.)

Reply:
Added not only a comment, but also a Chef `log` resource in case state changes on an instance already using this cookbook.